### PR TITLE
[Opt] speed up data load for duplicate table no-vec-load

### PR DESCRIPTION
**Pull request by zbtzbtzbt**
# Proposed changes

issue number：https://github.com/apache/incubator-doris/issues/8522

## Problem Summary:

similar pr：https://github.com/apache/incubator-doris/pull/7967

via `perf` command, I find data loading mainly cost is from `insert()` funtion's `get_type_info()`,
so I made some targeted performance tuning.

@HappenLee @yangzhg @spaces-X @zenoyang 

thanks.

### CODE REVIEW

File: be/src/olap/olap_common.h [hunk 1]
```diff
@@ -143,7 +142,8 @@ enum FieldType {
     OLAP_FIELD_TYPE_HLL = 23,
     OLAP_FIELD_TYPE_BOOL = 24,
     OLAP_FIELD_TYPE_OBJECT = 25,
-    OLAP_FIELD_TYPE_STRING = 26
+    OLAP_FIELD_TYPE_STRING = 26,
+    OLAP_FIELD_TYPE_SIZE = 27 // Size of FieldType
```
Comment (morningman) [inline be/src/olap/olap_common.h:146]: And comment to notice that this must be the last field of `FieldType`

Comment (zbtzbtzbt) [reply be/src/olap/olap_common.h]: > And comment to notice that this must be the last field of `FieldType`

done

**Comment by zbtzbtzbt** [comment]
Exciting news, I just tested it vai 100G SSB data and find  **almost faster X1**！

- 100G SSB data for loading
- 1 FE
- 3 BE

before: loading 100G needs **6min55s**；
after: loading 100G needs  **3min45s** 

I will confirm again and again
and if it's true, I will open an issue to record detailed experimental data tomorrow.

@HappenLee @morningman

**Comment by kpfly** [comment]
Good Job!

zbtzbtzbt ***@***.***> 于2022年3月17日周四 20:33写道：

File: be/src/olap/types.cpp [hunk 2]
```diff
@@ -85,7 +82,11 @@ ScalarTypeInfoResolver::ScalarTypeInfoResolver() {
     add_mapping<OLAP_FIELD_TYPE_OBJECT>();
 }
 
-ScalarTypeInfoResolver::~ScalarTypeInfoResolver() {}
+ScalarTypeInfoResolver::~ScalarTypeInfoResolver() {
```
[the remark this answers is not in the document]
Comment (zbtzbtzbt) [reply be/src/olap/types.cpp:85]: ok

File: be/src/olap/olap_common.h [hunk 3]
```diff
@@ -143,7 +142,8 @@ enum FieldType {
     OLAP_FIELD_TYPE_HLL = 23,
     OLAP_FIELD_TYPE_BOOL = 24,
     OLAP_FIELD_TYPE_OBJECT = 25,
-    OLAP_FIELD_TYPE_STRING = 26
+    OLAP_FIELD_TYPE_STRING = 26,
+    OLAP_FIELD_TYPE_SIZE = 27 // Size of FieldType, Notice: This must be the last field of FieldType
```
Comment (spaces-X) [inline be/src/olap/olap_common.h:146]: May be we should enlarge this value to make sure it can be alway at the last field in enum.

**Comment by cambyzju** [comment]
every cell insert always need to call get_type_info once, it doesn't sound reasonable.
Should we improve it by cache TypeInfo?

**Comment by zbtzbtzbt** [comment]
> every cell insert always need to call get_type_info once, it doesn't sound reasonable. Should we improve it by cache TypeInfo?

yes, There is still a lot of room for improvement

File: be/src/olap/types.cpp [hunk 4]
```diff
@@ -44,27 +44,24 @@ class ScalarTypeInfoResolver {
     DECLARE_SINGLETON(ScalarTypeInfoResolver);
 
 public:
-    std::shared_ptr<const TypeInfo> get_type_info(const FieldType t) {
-        auto pair = _scalar_type_mapping.find(t);
-        DCHECK(pair != _scalar_type_mapping.end()) << "Bad field type: " << t;
-        return pair->second;
+    std::shared_ptr<const TypeInfo> get_type_info(const FieldType t) const {
+        return _scalar_type_mapping[t];
```
Comment (caiconghui) [inline be/src/olap/types.cpp:48]: how about use template here？
so we don't keep OLAP_FIELD_TYPE_SIZE and _scalar_type_mapping？

Comment (zbtzbtzbt) [reply be/src/olap/types.cpp:48]: Sounds like it can be returned at compile time?

File: be/src/olap/olap_common.h (hunk 3 above)
Comment (zbtzbtzbt) [reply be/src/olap/olap_common.h:146]: ok,i try to delete this ugly param

File: be/src/olap/types.cpp (hunk 4 above)
Comment (caiconghui) [reply be/src/olap/types.cpp:48]: It should be that template functions are specialized for each type at compile time

Comment (wangbo) [reply be/src/olap/types.cpp:48]: >It should be that template functions are specialized for each type at compile time
+1

**Comment by zbtzbtzbt** [comment]
I tried to use template, but get an error when compiling, I have also tried different ways of writing,
 all of which are the same error, can someone help me? thanks
```
/incubator-doris/be/src/olap/aggregate_func.h:106: undefined reference to 
`std::shared_ptr<doris::TypeInfo const> doris::get_scalar_type_info_by_template<(doris::FieldType)9>()'
src/olap/libOlap.a(aggregate_func.cpp.o): In function `doris::BaseAggregateFuncs<(doris::FieldType)10, (doris::FieldType)22>::init(doris::RowCursorCell*, char const*, bool, doris::MemPool*, doris::ObjectPool*)':
/incubator-doris/be/src/olap/aggregate_func.h:106: undefined reference to `std::shared_ptr<doris::TypeInfo const> doris::get_scalar_type_info_by_template<(doris::FieldType)10>()'
src/olap/libOlap.a(aggregate_func.cpp.o): In function `doris::BaseAggregateFuncs<(doris::FieldType)11, (doris::FieldType)22>::init(doris::RowCursorCell*, char const*, bool, doris::MemPool*, doris::ObjectPool*)':
/incubator-doris/be/src/olap/aggregate_func.h:106: undefined reference to `std::shared_ptr<doris::TypeInfo const> doris::get_scalar_type_info_by_template<(doris::FieldType)11>()'
collect2: error: ld returned 1 exit status
```

**Comment by zbtzbtzbt** [comment]
> I tried to use template, but get an error when compiling, I have also tried different ways of writing, all of which are the same error, can someone help me? thanks
> 
use static and constexpr, The following method is also the same error:
 ```
template <FieldType field_type>
std::shared_ptr<const TypeInfo> get_scalar_type_info_by_template() {
    static constexpr TypeTraits<field_type> traits;
    static constexpr ScalarTypeInfo p = ScalarTypeInfo(traits);
    return std::shared_ptr<const TypeInfo>(&p);
}
```

**Comment by morningman** [comment]
I test it in my own env. 1 FE, 1 BE.
```
CREATE TABLE `t2_0` (
  `k` bigint(20) NOT NULL COMMENT "",
  `v0` bigint(20) NOT NULL COMMENT "",
  `v1` bigint(20) NOT NULL COMMENT "",
  `v2` bigint(20) NOT NULL COMMENT "",
  `v3` bigint(20) NOT NULL COMMENT "",
  `v4` bigint(20) NOT NULL COMMENT "",
  `v5` bigint(20) NOT NULL COMMENT "",
  `v6` bigint(20) NOT NULL COMMENT "",
  `v7` bigint(20) NOT NULL COMMENT "",
  `vfull` bigint(20) NOT NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`k`, `v0`, `v1`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k`) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"in_memory" = "false",
"storage_format" = "V2"
)
```

with data like this:
```
1|1|1|1|1|1|1|1|1|1|
1|1|1|1|1|1|1|1|1|2|
1|1|1|1|1|1|1|1|1|3|
1|1|1|1|1|1|1|1|1|4|
1|1|1|1|1|1|1|1|1|5|
1|1|1|1|1|1|1|1|1|6|
1|1|1|1|1|1|1|1|1|7|
1|1|1|1|1|1|1|1|1|8|
1|1|1|1|1|1|1|1|1|9|
1|1|1|1|1|1|1|1|1|10|
```

and total 104857600 rows.

Before: 104474 ms
After: 76021 ms
Improve: 27%

**Comment by zbtzbtzbt** [comment]
> I test it in my own env. 1 FE, 1 BE.
> and total 104857600 rows.
> Before: 104474 ms After: 76021 ms Improve: 27%

I'm glad it works, although not so obvious as some of my test.
based on code review comments, I'm trying to use templates, 
which means the type is known at compile time, and it will faster.

File: be/src/olap/types.h [hunk 5]
```diff
@@ -1193,6 +1208,23 @@ struct TypeTraits : public FieldTypeTraits<field_type> {
     static const int32_t size = sizeof(CppType);
 };
 
+// get type at compile time for performance
+// ---- Version_1 ----
+// template <FieldType field_type>
```
Comment (morningman) [inline be/src/olap/types.h:1213]: Remove unused code

File: be/src/olap/types.h [hunk 6]
```diff
@@ -1193,6 +1208,23 @@ struct TypeTraits : public FieldTypeTraits<field_type> {
...
+//     static auto obj = ScalarTypeInfo(traits);
+//     static constexpr ScalarTypeInfo* p = &obj;
+//     return (TypeInfo*)(p);
+// }
+
+// Version_2
+template <FieldType field_type>
+std::shared_ptr<const TypeInfo> get_scalar_type_info_by_template() {
+    static constexpr TypeTraits<field_type> traits;
+    return std::shared_ptr<const TypeInfo>(new ScalarTypeInfo(traits));
```
Comment (caiconghui) [inline be/src/olap/types.h:1225]: how about define static TypeInfo object and return reference. now， it seems we need to
create ScalarTypeInfo every function call

File: be/src/olap/types.h [hunk 7]
```diff
@@ -1193,6 +1208,23 @@ struct TypeTraits : public FieldTypeTraits<field_type> {
...
+// extern TypeInfo* get_scalar_type_info_by_template(){
+//     static constexpr TypeTraits<field_type> traits;
+//     static auto obj = ScalarTypeInfo(traits);
+//     static constexpr ScalarTypeInfo* p = &obj;
+//     return (TypeInfo*)(p);
+// }
+
+// Version_2
+template <FieldType field_type>
+std::shared_ptr<const TypeInfo> get_scalar_type_info_by_template() {
```
[the remark this answers is not in the document]
Comment (zbtzbtzbt) [reply be/src/olap/types.h:1223]: ok

**Comment by zbtzbtzbt** [comment]
conflict solved.

**Review by caiconghui — APPROVED** [review]
LGTM

**Pull request merged**